diff --git a/src/FSAL/commonlib.c b/src/FSAL/commonlib.c
index f53d170..a53b11b 100644
--- a/src/FSAL/commonlib.c
+++ b/src/FSAL/commonlib.c
@@ -962,6 +962,22 @@ static bool posix_get_fsid(struct fsal_filesystem *fs)
 			"stat_fs of %s resulted in error %s(%d)",
 			fs->path, strerror(errno), errno);
 
+	/*
+	There is an issue in docker where VOLUMES all share the same device node major and minor number.
+	As a result, only the first one is added to the list of volumes. Using this patch we change the
+	behavior to only look for a single volume mounted at /opt/target. This is of course not a valid
+	patch for the mainstream Ganesha version but serve our purpose in the cookie environment. The
+	correct way to do that is to add an extra distinctive parameter to identify a filesystem in
+	addition to the major and minor number, or to have a fix in docker.
+	*/
+	if (strcmp(fs->path, "/opt/target") != 0) {
+		LogCrit(COMPONENT_FSAL,
+			"COOKIE: ignore mount point %s that is not '/opt/target' to avoid conflict",
+			fs->path
+		);
+		return false;
+	}
+
 #if __FreeBSD__
 	fs->namelen = stat_fs.f_namemax;
 #else
